File: packages/domain/src/revisions.ts
import {
  ClearanceItemSchema,
  type AuditEvent,
  type ClearanceItem,
  type CutVersion,
  type Project,
  type ProjectRevision,
  type RevisionApplyResult,
  type RevisionChange,
  type ScriptVersion,
} from "@clearcut/contracts";

import { applyDisposition, DomainError, hasHumanDecision } from "./status";

export class RevisionComparisonError extends DomainError {
  constructor(message: string) {
    super("revision_comparison_error", message);
    this.name = "RevisionComparisonError";
  }
}

export class RevisionConflictError extends DomainError {
  constructor(message = "The active revision changed. Refresh and compare again.") {
    super("revision_conflict", message);
    this.name = "RevisionConflictError";
  }
}

function normalize(value: string): string {
  return value.toLocaleLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
}

function scenes(item: ClearanceItem): Set<number> {
  return new Set(item.script_references.map((reference) => reference.scene_index));
}

function times(item: ClearanceItem): Array<[number, number]> {
  return item.cut_detections.map((detection) => [
    Math.round(detection.timecode.start * 10) / 10,
    Math.round(detection.timecode.end * 10) / 10,
  ]);
}

function equalSets(left: Set<number>, right: Set<number>): boolean {
  return left.size === right.size && [...left].every((value) => right.has(value));
}

function equalTimes(left: Array<[number, number]>, right: Array<[number, number]>): boolean {
  return left.length === right.length && left.every((value, index) => value[0] === right[index]?.[0] && value[1] === right[index]?.[1]);
}

function sourceProximity(before: ClearanceItem, after: ClearanceItem): boolean {
  const beforeScenes = scenes(before);
  const afterScenes = scenes(after);
  if (beforeScenes.size && afterScenes.size && [...beforeScenes].some((scene) => afterScenes.has(scene))) {
    return true;
  }
  return times(before).some(([oldStart]) => times(after).some(([newStart]) => Math.abs(oldStart - newStart) <= 8));
}

function anchorsChanged(before: ClearanceItem, after: ClearanceItem): boolean {
  const beforeScenes = scenes(before);
  const afterScenes = scenes(after);
  if ((beforeScenes.size || afterScenes.size) && !equalSets(beforeScenes, afterScenes)) return true;
  const beforeTimes = times(before);
  const afterTimes = times(after);
  if ((beforeTimes.length || afterTimes.length) && !equalTimes(beforeTimes, afterTimes)) return true;
  return before.provenance !== after.provenance;
}

function revisions(project: Project): ProjectRevision[] {
  if (!project.revisions) project.revisions = [];
  return project.revisions;
}

function activeRevision(project: Project): ProjectRevision {
  const all = revisions(project);
  const active = project.active_revision_id
    ? all.find((revision) => revision.id === project.active_revision_id)
    : all.at(-1);
  if (!active) throw new RevisionComparisonError("The project has no active revision snapshot.");
  return active;
}

export function ensureInitialRevision(project: Project): ProjectRevision {
  const all = revisions(project);
  if (all.length) {
    if (!project.active_revision_id) project.active_revision_id = all.at(-1)?.id ?? null;
    return activeRevision(project);
  }
  const now = new Date().toISOString();
  const revision: ProjectRevision = {
    id: `revision_${crypto.randomUUID()}`,
    sequence: 1,
    script: structuredClone(project.script),
    cut: structuredClone(project.cut),
    state: "applied",
    items: structuredClone(project.items),
    changes: [],
    predecessor_id: null,
    created_at: now,
    applied_at: now,
    error: null,
  };
  all.push(revision);
  project.active_revision_id = revision.id;
  return revision;
}

function findMatches(previousItems: ClearanceItem[], candidates: ClearanceItem[]) {
  const unmatched = new Map(previousItems.map((item) => [item.id, item]));
  const byStable = new Map(previousItems.map((item) => [item.stable_item_id, item]));
  const matches = new Map<string, { item: ClearanceItem; basis: string }>();

  for (const candidate of candidates) {
    const explicit = byStable.get(candidate.stable_item_id);
    if (explicit && unmatched.has(explicit.id)) {
      matches.set(candidate.id, { item: explicit, basis: "explicit stable identity" });
      unmatched.delete(explicit.id);
    }
  }
  for (const candidate of candidates) {
    if (matches.has(candidate.id)) continue;
    const exact = [...unmatched.values()].find((item) =>
      normalize(item.name) === normalize(candidate.name) && item.category === candidate.category
    );
    if (exact) {
      matches.set(candidate.id, { item: exact, basis: "normalized category and name" });
      unmatched.delete(exact.id);
    }
  }
  for (const candidate of candidates) {
    if (matches.has(candidate.id)) continue;
    const nearby = [...unmatched.values()].filter((item) =>
      item.category === candidate.category && sourceProximity(item, candidate)
    );
    if (nearby.length === 1) {
      matches.set(candidate.id, { item: nearby[0], basis: "category and source proximity" });
      unmatched.delete(nearby[0].id);
    }
  }
  return { matches, removed: unmatched };
}

export function compareRevisions(previous: ProjectRevision, candidateItems: ClearanceItem[]): RevisionChange[] {
  const parsed = candidateItems.map((candidate) => ClearanceItemSchema.safeParse(candidate));
  if (parsed.some((result) => !result.success)) {
    throw new RevisionComparisonError("Every revision candidate must be a clearance item.");
  }
  const candidates = parsed.map((result) => result.success ? result.data : neverCandidate());
  if (new Set(candidates.map((item) => item.id)).size !== candidates.length) {
    throw new RevisionComparisonError("Revision candidate item IDs must be unique.");
  }
  const { matches, removed } = findMatches(previous.items, candidates);
  const changes: RevisionChange[] = [];
  for (const candidate of candidates) {
    const match = matches.get(candidate.id);
    if (!match) {
      changes.push({
        kind: "added",
        stable_item_id: candidate.stable_item_id,
        item_name: candidate.name,
        before_item_id: null,
        after_item_id: candidate.id,
        explanation: "New clearance element in the candidate revision.",
        match_basis: "no safe predecessor match",
        previous_status: null,
      });
      continue;
    }
    const identityChanged = normalize(match.item.name) !== normalize(candidate.name)
      || match.item.category !== candidate.category;
    const stale = hasHumanDecision(match.item) && anchorsChanged(match.item, candidate);
    const kind = stale ? "decision_stale" : identityChanged ? "materially_changed" : "unchanged";
    const explanation = kind === "unchanged"
      ? "Identity and recorded source anchors are unchanged."
      : kind === "materially_changed"
        ? "The element became materially more specific or changed identity."
        : "A prior human disposition predates changed source anchors or use.";
    changes.push({
      kind,
      stable_item_id: match.item.stable_item_id,
      item_name: candidate.name,
      before_item_id: match.item.id,
      after_item_id: candidate.id,
      explanation,
      match_basis: match.basis,
      previous_status: match.item.workflow_status,
    });
  }
  for (const removedItem of removed.values()) {
    changes.push({
      kind: "removed",
      stable_item_id: removedItem.stable_item_id,
      item_name: removedItem.name,
      before_item_id: removedItem.id,
      after_item_id: null,
      explanation: "The element is absent from the candidate revision.",
      match_basis: "no safe candidate match",
      previous_status: removedItem.workflow_status,
    });
  }
  return changes;
}

function neverCandidate(): never {
  throw new RevisionComparisonError("Every revision candidate must be a clearance item.");
}

function carryRecord(before: ClearanceItem, after: ClearanceItem): ClearanceItem {
  return {
    ...structuredClone(before),
    id: after.id,
    stable_item_id: before.stable_item_id,
    name: after.name,
    category: after.category,
    description: after.description,
    provenance: after.provenance,
    source_version: after.source_version,
    script_references: structuredClone(after.script_references),
    cut_detections: structuredClone(after.cut_detections),
    detection_confidence: after.detection_confidence,
    research_priority: after.research_priority,
    production_impact: after.production_impact,
  };
}

export function buildCandidateRevision(
  project: Project,
  candidateItems: ClearanceItem[],
  options: { script?: ScriptVersion | null; cut?: CutVersion | null; revisionId?: string } = {},
): ProjectRevision {
  const previous = activeRevision(project);
  const changes = compareRevisions(previous, candidateItems);
  const byAfter = new Map(changes.filter((change) => change.after_item_id).map((change) => [change.after_item_id, change]));
  const previousById = new Map(previous.items.map((item) => [item.id, item]));
  const items = candidateItems.map((candidate) => {
    const change = byAfter.get(candidate.id);
    if (!change) throw new RevisionComparisonError("Candidate comparison is incomplete.");
    if (!change.before_item_id) return structuredClone(candidate);
    const before = previousById.get(change.before_item_id);
    if (!before) throw new RevisionComparisonError("Candidate predecessor is missing.");
    return carryRecord(before, candidate);
  });
  return {
    id: options.revisionId ?? `revision_${crypto.randomUUID()}`,
    sequence: Math.max(0, ...revisions(project).map((revision) => revision.sequence)) + 1,
    script: structuredClone(options.script === undefined ? project.script : options.script),
    cut: structuredClone(options.cut === undefined ? project.cut : options.cut),
    state: "ready",
    items,
    changes,
    predecessor_id: previous.id,
    created_at: new Date().toISOString(),
    applied_at: null,
    error: null,
  };
}

export function applyRevision(
  project: Project,
  revisionId: string,
  predecessorId: string | null,
): RevisionApplyResult {
  const revision = revisions(project).find((candidate) => candidate.id === revisionId);
  if (!revision) throw new RevisionComparisonError("Revision not found.");
  if (revision.state === "applied" && project.active_revision_id === revision.id) {
    return { revision, project, already_applied: true };
  }
  if (revision.state !== "ready") {
    throw new RevisionComparisonError("Only a ready comparison can be applied.");
  }
  if (predecessorId !== revision.predecessor_id || project.active_revision_id !== revision.predecessor_id) {
    throw new RevisionConflictError();
  }

  const promoted = structuredClone(revision.items);
  const byStable = new Map(promoted.map((item) => [item.stable_item_id, item]));
  for (const change of revision.changes) {
    if (change.kind !== "materially_changed" && change.kind !== "decision_stale") continue;
    const target = byStable.get(change.stable_item_id);
    if (!target) continue;
    const reopened = applyDisposition(target, {
      actor: "system",
      status: "reopened_by_revision",
      rationale: change.explanation,
      detail: {
        revision_id: revision.id,
        previous_status: change.previous_status,
        before_item_id: change.before_item_id,
        after_item_id: change.after_item_id,
      },
    });
    Object.assign(target, reopened);
  }

  project.items = promoted;
  project.script = structuredClone(revision.script);
  project.cut = structuredClone(revision.cut);
  project.script_history ??= [];
  project.cut_history ??= [];
  if (project.script && !project.script_history.some((version) => version.id === project.script?.id)) {
    project.script_history.push(structuredClone(project.script));
  }
  if (project.cut && !project.cut_history.some((version) => version.id === project.cut?.id)) {
    project.cut_history.push(structuredClone(project.cut));
  }
  const now = new Date().toISOString();
  project.active_revision_id = revision.id;
  revision.state = "applied";
  revision.applied_at = now;
  project.updated_at = now;
  const event: AuditEvent = {
    id: `evt_${crypto.randomUUID()}`,
    at: now,
    actor: "coordinator",
    actor_name: "",
    action: "revision_applied",
    from_status: null,
    to_status: null,
    rationale: `Applied revision ${revision.sequence} after reviewing its comparison.`,
    source_version: project.cut?.label ?? project.script?.label ?? "",
    detail: {
      revision_id: revision.id,
      predecessor_id: revision.predecessor_id,
      reopened: revision.changes.filter((change) => change.kind === "materially_changed" || change.kind === "decision_stale").length,
    },
  };
  project.audit_events.push(event);
  return { revision, project, already_applied: false };
}

File: backend/app/revisions.py
"""Immutable candidate revisions and deterministic clearance-state comparison."""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Iterable

from .models import (
    ClearanceItem,
    CutVersion,
    Project,
    ProjectRevision,
    RevisionChange,
    ScriptVersion,
)


class RevisionComparisonError(ValueError):
    """Raised when a candidate cannot be compared without unsafe assumptions."""


class RevisionConflictError(RuntimeError):
    """Raised when a candidate no longer follows the active revision."""


def _normal(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", " ", value.casefold()).strip()


def _scenes(item: ClearanceItem) -> set[int]:
    return {reference.scene_index for reference in item.script_references}


def _times(item: ClearanceItem) -> list[tuple[float, float]]:
    return [
        (round(detection.timecode.start, 1), round(detection.timecode.end, 1))
        for detection in item.cut_detections
    ]


def _source_proximity(before: ClearanceItem, after: ClearanceItem) -> bool:
    before_scenes = _scenes(before)
    after_scenes = _scenes(after)
    if before_scenes and after_scenes and before_scenes & after_scenes:
        return True
    before_times = _times(before)
    after_times = _times(after)
    return any(
        abs(old_start - new_start) <= 8
        for old_start, _ in before_times
        for new_start, _ in after_times
    )


def _anchors_changed(before: ClearanceItem, after: ClearanceItem) -> bool:
    if _scenes(before) or _scenes(after):
        if _scenes(before) != _scenes(after):
            return True
    if _times(before) or _times(after):
        if _times(before) != _times(after):
            return True
    return before.provenance != after.provenance


def _active_revision(project: Project) -> ProjectRevision:
    if project.active_revision_id:
        active = next(
            (revision for revision in project.revisions if revision.id == project.active_revision_id),
            None,
        )
        if active is not None:
            return active
    if project.revisions:
        return project.revisions[-1]
    raise RevisionComparisonError("The project has no active revision snapshot.")


def ensure_initial_revision(project: Project) -> ProjectRevision:
    """Snapshot the first completed project exactly once."""
    if project.revisions:
        if not project.active_revision_id:
            project.active_revision_id = project.revisions[-1].id
        return _active_revision(project)
    revision = ProjectRevision(
        sequence=1,
        script=project.script.model_copy(deep=True) if project.script else None,
        cut=project.cut.model_copy(deep=True) if project.cut else None,
        state="applied",
        items=[item.model_copy(deep=True) for item in project.items],
        changes=[],
        applied_at=datetime.now(timezone.utc).isoformat(),
    )
    project.revisions.append(revision)
    project.active_revision_id = revision.id
    return revision


def _find_matches(
    previous_items: list[ClearanceItem],
    candidate_items: list[ClearanceItem],
) -> tuple[dict[str, tuple[ClearanceItem, str]], set[str]]:
    unmatched = {item.id: item for item in previous_items}
    by_stable = {item.stable_item_id: item for item in previous_items}
    matches: dict[str, tuple[ClearanceItem, str]] = {}

    for candidate in candidate_items:
        explicit = by_stable.get(candidate.stable_item_id)
        if explicit and explicit.id in unmatched:
            matches[candidate.id] = (explicit, "explicit stable identity")
            unmatched.pop(explicit.id)

    for candidate in candidate_items:
        if candidate.id in matches:
            continue
        exact = next(
            (
                item
                for item in unmatched.values()
                if _normal(item.name) == _normal(candidate.name)
                and item.category == candidate.category
            ),
            None,
        )
        if exact:
            matches[candidate.id] = (exact, "normalized category and name")
            unmatched.pop(exact.id)

    for candidate in candidate_items:
        if candidate.id in matches:
            continue
        nearby = [
            item
            for item in unmatched.values()
            if item.category == candidate.category and _source_proximity(item, candidate)
        ]
        if len(nearby) == 1:
            matches[candidate.id] = (nearby[0], "category and source proximity")
            unmatched.pop(nearby[0].id)

    return matches, set(unmatched)


def compare_revisions(
    previous: ProjectRevision,
    candidate_items: Iterable[ClearanceItem],
) -> list[RevisionChange]:
    candidates = list(candidate_items)
    if any(not isinstance(item, ClearanceItem) for item in candidates):
        raise RevisionComparisonError("Every revision candidate must be a clearance item.")
    if len({item.id for item in candidates}) != len(candidates):
        raise RevisionComparisonError("Revision candidate item IDs must be unique.")

    matches, removed_ids = _find_matches(previous.items, candidates)
    changes: list[RevisionChange] = []
    for candidate in candidates:
        match = matches.get(candidate.id)
        if match is None:
            changes.append(
                RevisionChange(
                    kind="added",
                    stable_item_id=candidate.stable_item_id,
                    item_name=candidate.name,
                    after_item_id=candidate.id,
                    explanation="New clearance element in the candidate revision.",
                    match_basis="no safe predecessor match",
                )
            )
            continue
        before, basis = match
        identity_changed = (
            _normal(before.name) != _normal(candidate.name)
            or before.category != candidate.category
        )
        stale = before.has_human_decision and _anchors_changed(before, candidate)
        kind = "decision_stale" if stale else "materially_changed" if identity_changed else "unchanged"
        explanation = {
            "unchanged": "Identity and recorded source anchors are unchanged.",
            "materially_changed": "The element became materially more specific or changed identity.",
            "decision_stale": "A prior human disposition predates changed source anchors or use.",
        }[kind]
        changes.append(
            RevisionChange(
                kind=kind,
                stable_item_id=before.stable_item_id,
                item_name=candidate.name,
                before_item_id=before.id,
                after_item_id=candidate.id,
                explanation=explanation,
                match_basis=basis,
                previous_status=before.workflow_status,
            )
        )
    for item in previous.items:
        if item.id in removed_ids:
            changes.append(
                RevisionChange(
                    kind="removed",
                    stable_item_id=item.stable_item_id,
                    item_name=item.name,
                    before_item_id=item.id,
                    explanation="The element is absent from the candidate revision.",
                    match_basis="no safe candidate match",
                    previous_status=item.workflow_status,
                )
            )
    return changes


def _carry_record(before: ClearanceItem, after: ClearanceItem) -> ClearanceItem:
    retained = before.model_copy(deep=True)
    retained.id = after.id
    retained.stable_item_id = before.stable_item_id
    retained.name = after.name
    retained.category = after.category
    retained.description = after.description
    retained.provenance = after.provenance
    retained.source_version = after.source_version
    retained.script_references = after.script_references
    retained.cut_detections = after.cut_detections
    retained.detection_confidence = after.detection_confidence
    retained.research_priority = after.research_priority
    retained.production_impact = after.production_impact
    return retained


def build_candidate_revision(
    project: Project,
    candidate_items: Iterable[ClearanceItem],
    *,
    script: ScriptVersion | None = None,
    cut: CutVersion | None = None,
    revision_id: str | None = None,
) -> ProjectRevision:
    """Build a detached ready snapshot; never mutates ``project``."""
    candidates = list(candidate_items)
    if any(not isinstance(item, ClearanceItem) for item in candidates):
        raise RevisionComparisonError("Every revision candidate must be a clearance item.")
    previous = _active_revision(project)
    changes = compare_revisions(previous, candidates)
    by_after = {change.after_item_id: change for change in changes if change.after_item_id}
    previous_by_id = {item.id: item for item in previous.items}
    items: list[ClearanceItem] = []
    for candidate in candidates:
        candidate_copy = candidate.model_copy(deep=True)
        change = by_after[candidate.id]
        if change.before_item_id:
            candidate_copy = _carry_record(previous_by_id[change.before_item_id], candidate_copy)
        items.append(candidate_copy)
    values = {
        "sequence": max((revision.sequence for revision in project.revisions), default=0) + 1,
        "script": (script or project.script).model_copy(deep=True) if (script or project.script) else None,
        "cut": (cut or project.cut).model_copy(deep=True) if (cut or project.cut) else None,
        "state": "ready",
        "items": items,
        "changes": changes,
        "predecessor_id": previous.id,
    }
    if revision_id:
        values["id"] = revision_id
    return ProjectRevision(**values)


def apply_revision(
    project: Project,
    revision_id: str,
    predecessor_id: str | None,
) -> tuple[ProjectRevision, bool]:
    revision = next((candidate for candidate in project.revisions if candidate.id == revision_id), None)
    if revision is None:
        raise RevisionComparisonError("Revision not found.")
    if revision.state == "applied" and project.active_revision_id == revision.id:
        return revision, True
    if revision.state != "ready":
        raise RevisionComparisonError("Only a ready comparison can be applied.")
    if predecessor_id != revision.predecessor_id or project.active_revision_id != revision.predecessor_id:
        raise RevisionConflictError("The active revision changed. Refresh and compare again.")

    promoted = [item.model_copy(deep=True) for item in revision.items]
    by_stable = {item.stable_item_id: item for item in promoted}
    for change in revision.changes:
        if change.kind not in {"materially_changed", "decision_stale"}:
            continue
        item = by_stable.get(change.stable_item_id)
        if item is None:
            continue
        item.transition(
            "reopened_by_revision",
            actor="system",
            rationale=change.explanation,
            detail={
                "revision_id": revision.id,
                "previous_status": change.previous_status,
                "before_item_id": change.before_item_id,
                "after_item_id": change.after_item_id,
            },
        )

    project.items = promoted
    project.script = revision.script.model_copy(deep=True) if revision.script else None
    project.cut = revision.cut.model_copy(deep=True) if revision.cut else None
    if project.script and all(version.id != project.script.id for version in project.script_history):
        project.script_history.append(project.script)
    if project.cut and all(version.id != project.cut.id for version in project.cut_history):
        project.cut_history.append(project.cut)
    project.active_revision_id = revision.id
    revision.state = "applied"
    revision.applied_at = datetime.now(timezone.utc).isoformat()
    project.updated_at = revision.applied_at
    project.log(
        "revision_applied",
        actor="coordinator",
        rationale=f"Applied revision {revision.sequence} after reviewing its comparison.",
        detail={
            "revision_id": revision.id,
            "predecessor_id": revision.predecessor_id,
            "reopened": sum(
                change.kind in {"materially_changed", "decision_stale"}
                for change in revision.changes
            ),
        },
    )
    return revision, False
